net/src/ipv4/prefix.rs

```diff
@@ -0,0 +1,665 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+#[cfg(any(test, feature = "bolero"))]
+#[allow(unused_imports)] // re-export
+pub use contract::*;
+use ipnet::{AddrParseError, Ipv4Net};
+use prefix_trie::Prefix;
+use serde::{Deserialize, Serialize};
+use std::fmt::{Debug, Display, Formatter};
+use std::net::Ipv4Addr;
+use std::str::FromStr;
+use tracing::{debug, warn};
+
+/// An `Ipv4Addr` with a mask describing a network in CIDR notation.
+///
+/// Note that unlike [`Ipv4Net`] from the `ipnet` crate, this type ensures that only network bits
+/// are set in the address.
+#[derive(
+    Default, Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize,
+)]
+#[repr(transparent)]
+#[serde(transparent)]
+pub struct Ipv4Prefix(Ipv4Net);
+
+/// A checked type describing the values 0 to 32, which constitute all legal prefix lengths for
+/// Ipv4 addresses.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[repr(transparent)]
+#[serde(try_from = "u8", into = "u8")]
+pub struct Ipv4PrefixLen(u8);
+
+/// An error indicating that an invalid prefix length was provided.
+#[derive(Debug, thiserror::Error)]
+#[repr(transparent)]
+pub enum InvalidIpv4PrefixLength {
+    /// The provided prefix is too long to form a legal [`Ipv4PrefixLen`]
+    #[error("invalid prefix length {0} is invalid, max is {MAX}", MAX = Ipv4PrefixLen::MAX_LEN)]
+    TooLong(u8),
+}
+
+/// An error indicating that an invalid network was provided.
+#[derive(Debug, thiserror::Error)]
+pub enum InvalidIpv4Network {
+    /// The provided network description contains set non-network bits
+    #[error("Address {0}/{1} contains non network bits")]
+    AddressContainsNonNetworkBits(Ipv4Addr, Ipv4PrefixLen),
+    /// The provided prefix length is invalid
+    #[error(transparent)]
+    InvalidPrefix(InvalidIpv4PrefixLength),
+}
+
+/// An error indicating that an invalid network was provided.
+#[derive(Debug, thiserror::Error)]
+pub enum Ipv4PrefixParseError {
+    /// failure to interpret string as an ip and a prefix length
+    #[error(transparent)]
+    AddrParseError(AddrParseError),
+    /// invalid ip or prefix length
+    #[error(transparent)]
+    InvalidIpv4Network(InvalidIpv4Network),
+}
+
+impl Ipv4PrefixLen {
+    /// The largest possible prefix length for IPv4 (i.e. /32)
+    pub const MAX_LEN: u8 = 32;
+    /// The largest possible prefix length for IPv4 (i.e. /32)
```

Comment:
Nit: s/largest/longest/ ? (I find that “largest” is somewhat confusing in that it lead to believe the prefix with the “largest” size will be the “largest”, when it is in fact the smaller one.)

Reply:
agree

Reply:
fair enough